### PR TITLE
feat(compass-app-stores): introduce MongoDBInstancesManager for supporting workspace initialisation without DataService / MongoDBInstance COMPASS-7718

### PR DESCRIPTION
**Pull request by himanshusinghs**
## Description
This is first of the PR related to work in COMPASS-7718. It introduces MongoDBInstancesManager, an event emitter that also keep track of MongoDBInstances for connected connections.

MongoDBInstancesManager is suppose to help us in removing direct dependency on MongoDBInstance and DataService from plugins that are expected to be rendered even in the non-connected part of the application tree. These plugins for now are WorkspacesPlugin, Sidebar and all the global modals that the WorkspacesPlugin renders. Changes for the mentioned plugins will follow in upcoming PRs.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-app-stores/src/instances-manager.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    this.instances.set(connectionInfoId, instance);
+    this.emit(
+      MongoDBInstancesManagerEvents.InstanceCreated,
+      connectionInfoId,
+      instance
+    );
+    return instance;
+  }
+
+  getMongoDBInstance(connectionInfoId: ConnectionInfo['id']) {
```
Comment (gribnoysup) [inline packages/compass-app-stores/src/instances-manager.ts:45]: Small suggestion: to make it clear what id is being passed here without the need to read the implementation maybe we can update the method name

```suggestion
  getMongoDBInstanceForConnection(connectionInfoId: ConnectionInfo['id']) {
```

**Review by gribnoysup — APPROVED** [review]

**Review by kmruiz — APPROVED** [review]

File: packages/compass-app-stores/src/stores/instance-store.ts [hunk 2]
```diff
@@ -52,7 +52,13 @@ export function createInstancesStore(
...
 ) {
-  const instances: Record<ConnectionInfo['id'], MongoDBInstance> = {};
+  const instancesManager = new MongoDBInstancesManager();
+  connectionsManager.on(
+    ConnectionsManagerEvents.ConnectionDisconnected,
+    function (connectionInfoId) {
+      instancesManager.removeMongoDBInstanceForConnection(connectionInfoId);
+    }
+  );
   connectionsManager.on(
```
[the remark this answers is not in the document]
Comment (himanshusinghs) [reply packages/compass-app-stores/src/stores/instance-store.ts:62]: Yea that is correct. I missed this, thank you pointing this out.

File: packages/compass-app-stores/src/instances-manager.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import { EventEmitter } from 'events';
+import { type ConnectionInfo } from '@mongodb-js/connection-info';
+import {
+  MongoDBInstance,
+  type MongoDBInstanceProps,
+} from 'mongodb-instance-model';
+
+export enum MongoDBInstancesManagerEvents {
+  InstanceCreated = 'instance-started',
+  InstanceRemoved = 'instance-removed',
```
[the remark this answers is not in the document]
Comment (himanshusinghs) [reply packages/compass-app-stores/src/instances-manager.ts]: Addressed this in [bf9ad26](https://github.com/mongodb-js/compass/pull/5669/commits/bf9ad26d1477d2fdbccb2cca8db465a60bfc845e)

File: packages/compass-app-stores/src/stores/instance-store.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (himanshusinghs) [reply packages/compass-app-stores/src/stores/instance-store.ts]: Addressed this in [bf9ad26](https://github.com/mongodb-js/compass/pull/5669/commits/bf9ad26d1477d2fdbccb2cca8db465a60bfc845e)

**Review by addaleax — APPROVED** [review]

**Pull request merged**